### PR TITLE
Issue #32: extend proxy template to support target servers

### PR DESCRIPTION
**Pull request by JoelGauci**
Extend proxy template to support target servers:

- sackmesser used for deployment on Apigee X, hybrid and Edge
- Generate target server configuration using a json template (edge.template.json)

**Fixes:** #32 

- [x] I have run all the tests locally and they all pass.
- [x] I have followed the relevant style guide for my changes.

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

File: references/proxy-template/template-v1/package.json [hunk 1]
```diff
@@ -7,12 +7,13 @@
     "test": "test"
   },
   "scripts": {
+    "deployTargetServer": "npm i && npx apigeetool createTargetServer --targetServerName $TARGET_SERVER_NAME -u $APIGEE_USER -p $APIGEE_PASS -o $APIGEE_ORG -e $APIGEE_ENV --targetHost $TARGET_HOST --targetPort $TARGET_PORT $APIGEETOOL_DEPLOY_TARGETSERVER_SSLOPTION -V",
     "deploy": "npm i && npx apigeetool deployproxy -u $APIGEE_USER -p $APIGEE_PASS -o $APIGEE_ORG -e $APIGEE_ENV -n $npm_package_name -V",
```
Comment (danistrebel) [inline references/proxy-template/template-v1/package.json:11]: we should maybe rename this to `deployProxy` now to be precise. WDYT?

File: references/proxy-template/template-v1/apiproxy/proxies/default.xml [hunk 2]
```diff
@@ -75,6 +76,14 @@
 
     <!-- insert your flows here and check scopes -->
 
+    <Flow name="GET /headers">
```
Comment (danistrebel) [inline references/proxy-template/template-v1/apiproxy/proxies/default.xml:79]: The `/header` might be confusing in a generic usage of the generator.

How about just using `request.verb = "GET" and proxy.pathsuffix MatchesPath "/**"`?

**Review by danistrebel — COMMENTED** [review]
I really like the idea of also generating a Target Server for the proxy.
The thing I struggle with at the moment is the now introduced tight coupling to the httpbin service or the `/headers` endpoint to be precise. 
I get that to be able to test we need a path and that this depends on the characteristics of the target server.

I see two options:

1. Don't use apickli to test the target server but use static inspection as the OAS mock generator does.

2. Pass the path for the flow condition to the generator and remove the hardcoded dependency

File: references/proxy-template/template-v1/package.json (hunk 1 above)
Comment (JoelGauci) [reply references/proxy-template/template-v1/package.json:11]: Yes agree with you.
So "deploy" becomes "deployProxy", right ?

Comment (JoelGauci) [reply references/proxy-template/template-v1/package.json:11]: I have replaced **deploy** with **deployProxy**

File: references/proxy-template/template-v1/apiproxy/proxies/default.xml (hunk 2 above)
Comment (JoelGauci) [reply references/proxy-template/template-v1/apiproxy/proxies/default.xml:79]: Modification: the flow name and condition are dynamically set based on the the path section of TARGET_URL.
The default value of TARGET_URL is https://httpbin.org/get

File: references/proxy-template/pipeline.sh [hunk 3]
```diff
@@ -18,17 +18,37 @@ set -x
...
 
 # deploy all shared flows
 (cd "$SCRIPTPATH"/../common-shared-flows && sh deploy.sh all --apigeeapi)
 
+# set target server environment variables
+DEFAULT_TARGET_URL=https://httpbin.org/get
+export TARGET_URL="${TARGET_URL:-"$DEFAULT_TARGET_URL"}"
+
+# shellcheck disable=SC1091
+. ./set-targetserver-envs.sh
```
Comment (danistrebel) [inline references/proxy-template/pipeline.sh:36]: why is disabling SC1091 needed here? Can't you use the SCRIPTPATH instead?

File: references/proxy-template/generate-proxy.sh [hunk 4]
```diff
@@ -13,16 +13,15 @@
...
 
-
 # Ask for input parameters if they are not set
 
 [ -z "$PROXY" ]     && printf "Proxy Name: "    && read -r PROXY
 [ -z "$VERSION" ]   && printf "Proxy Version: " && read -r VERSION
 [ -z "$VHOST" ]     && printf "Virtual Host: "  && read -r VHOST
-[ -z "$TARGETURL" ] && printf "Target URL: "    && read -r TARGETURL
+[ -z "$TARGET_SERVER_NAME" ] && printf "Target Server Name: "    && read -r TARGET_SERVER_NAME
+[ -z "$TARGET_PATH" ] && printf "Target Path: "    && read -r TARGET_PATH
```
Comment (danistrebel) [inline references/proxy-template/generate-proxy.sh:24]: Not sure I fully understand the mechanics here. There is a TARGET_PATH variable set by the user explicitly here and one by splitting the TARGET_URL in the `set-targetserver-envs.sh` isn't one the proxy path suffix?

Comment (JoelGauci) [reply references/proxy-template/generate-proxy.sh:24]: Yes I agree Daniel, this is not clear

File: references/proxy-template/pipeline.sh (hunk 3 above)
Comment (JoelGauci) [reply references/proxy-template/pipeline.sh]: SCRIPTPATH added

File: references/proxy-template/README.md [hunk 5]
```diff
@@ -27,15 +27,20 @@ It also has the following features:
 
 ## Quick start
 
+    # Set environment variables to create target server and target path
+    source ./set-targetserver-envs.sh
```
Comment (danistrebel) [inline references/proxy-template/README.md:31]: can we do this in the `generate-proxy.sh`?

File: references/proxy-template/generate-proxy.sh [hunk 6]
```diff
@@ -38,11 +71,19 @@ sed -i.bak "s|@Proxy@|$PROXY|" ./"$PROXY"-"$VERSION"/test/features/step_definiti
...
 rm ./"$PROXY"-"$VERSION"/apiproxy/proxies/default.xml.bak
 rm ./"$PROXY"-"$VERSION"/package.json.bak
 rm ./"$PROXY"-"$VERSION"/test/features/step_definitions/init.js.bak
+rm ./"$PROXY"-"$VERSION"/test/features/TargetServer.feature.bak
 rm ./"$PROXY"-"$VERSION"/apiproxy/targets/default.xml.bak
 
+export APIGEETOOL_DEPLOY_TARGETSERVER_SSLOPTION
+export TARGET_HOST
+export TARGET_PORT
+export TARGET_SERVER_NAME
```
Comment (danistrebel) [inline references/proxy-template/generate-proxy.sh:87]: @JoelGauci  What's the purpose of exporting these here?

Comment (JoelGauci) [reply references/proxy-template/generate-proxy.sh:87]: @danistrebel : I need these variables to test if target server exists and to create it if it does not (using **npm run deployTargetServer**)

**Comment by OmidTahouri** [comment]
As raised by @danistrebel, it would be preferred to not rely on `apigeetool` for this and instead use tooling that makes this project compatible with Apigee X/hybrid.

File: references/proxy-template/pipeline.sh [hunk 7]
```diff
@@ -14,21 +14,65 @@
...
 # deploy all shared flows
-(cd "$SCRIPTPATH"/../common-shared-flows && sh deploy.sh all --apigeeapi)
+(cd "$SCRIPTPATH"/../common-shared-flows && sh deploy.sh all $apiversion)
+
+# set target URL that is used to configure a target server
+TARGET_URL="${TARGET_URL:-"$DEFAULT_TARGET_URL"}"
+
+# generate the proxy
+echo "[INFO] Creating Proxy template and Target Server configuration"
+PROXY="$PROXY" VERSION="$VERSION" VHOST="$VHOST" TARGET_URL="$TARGET_URL" "$SCRIPTPATH"/generate-proxy.sh $apiversion
```
Comment (OmidTahouri) [inline references/proxy-template/pipeline.sh:56]: No big deal, but I'm leaning more towards separate lines for readability

**Review by OmidTahouri — CHANGES_REQUESTED** [review]
Please just update the npm scripts in package.json before we merge this :) Thanks, Joel. Great contribution!

File: references/proxy-template/template-v1/package.json [hunk 8]
```diff
@@ -7,12 +7,13 @@
     "test": "test"
   },
   "scripts": {
-    "deploy": "npm i && npx apigeetool deployproxy -u $APIGEE_USER -p $APIGEE_PASS -o $APIGEE_ORG -e $APIGEE_ENV -n $npm_package_name -V",
+    "deployTargetServer": "npm i && npx apigeetool createTargetServer --targetServerName $TARGET_SERVER_NAME -u $APIGEE_USER -p $APIGEE_PASS -o $APIGEE_ORG -e $APIGEE_ENV --targetHost $TARGET_HOST --targetPort $TARGET_PORT $APIGEETOOL_DEPLOY_TARGETSERVER_SSLOPTION -V",
+    "deployProxy": "npm i && npx apigeetool deployproxy -u $APIGEE_USER -p $APIGEE_PASS -o $APIGEE_ORG -e $APIGEE_ENV -n $npm_package_name -V",
```
[the remark this answers is not in the document]
Comment (JoelGauci) [reply references/proxy-template/template-v1/package.json:11]: Yes Omid, we do not use these lines anymore

File: references/proxy-template/README.md [hunk 9]
```diff
@@ -14,28 +14,71 @@ The provided template uses Shared Flows for:
...
+  - The path that is set in `TARGET_URL` is used to configure a conditional flow
+on the proxy template. This flow is used to test the TargetServer
+  - The default value of `TARGET_URL` is `https://httpbin.org/headers`
 
 ## Dependencies
 
+- [Maven](https://maven.apache.org/)
 - [Common Shared Flows](../common-shared-flows)
 - [NodeJS](https://nodejs.org/en/) LTS version or above
 - Apigee Evaluation [Organization](https://login.apigee.com/sign__up)
```
[the remark this answers is not in the document]
Comment (JoelGauci) [reply references/proxy-template/README.md:32]: I remove the "Apigee Evaluation" line but keep the others

**Review by OmidTahouri — APPROVED** [review]
LGTM

File: references/proxy-template/pipeline.sh [hunk 10]
```diff
@@ -14,21 +14,79 @@
...
 
 set -e
-set -x
 
 SCRIPTPATH="$( cd "$(dirname "$0")" || exit >/dev/null 2>&1 ; pwd -P )"
 
+if [ -z "$1" ] || [ "$1" = "--apigeeapi" ];then
+    apiversion="--apigeeapi"
+    TEST_HOST="$APIGEE_ORG-$APIGEE_ENV.apigee.net"
+elif [ -z "$1" ] || [ "$1" = "--googleapi" ];then
```
Comment (danistrebel) [inline references/proxy-template/pipeline.sh:23]: this means that we're not running tests against X/hybird? I'd rather we test against both targets sequentially.

**Review by danistrebel — CHANGES_REQUESTED** [review]
soo close. One last thing, please lease check the pipeline.sh to validate that we're running tests against edge and x..

File: references/proxy-template/pipeline.sh (hunk 10 above)
Comment (JoelGauci) [reply references/proxy-template/pipeline.sh:23]: sure, I modify the pipeline script

File: references/proxy-template/pipeline.sh [hunk 11]
```diff
@@ -14,21 +14,81 @@
...
 SCRIPTPATH="$( cd "$(dirname "$0")" || exit >/dev/null 2>&1 ; pwd -P )"
 
+if [ -z "$1" ] || [ "$1" = "--apigeeapi" ];then
+    apiversion="--apigeeapi"
+    TEST_HOST="$APIGEE_ORG-$APIGEE_ENV.apigee.net"
+fi
+
+if [ -z "$1" ] || [ "$1" = "--googleapi" ];then
+    apiversion="--googleapi"
+    TEST_HOST="$APIGEE_X_HOSTNAME"
```
Comment (danistrebel) [inline references/proxy-template/pipeline.sh:27]: if $1 is not set then you'll still only run tests against one deployment option. It probably makes sense to duplicate the logic below more or create a deploy.sh file that you can use for both.

**Review by danistrebel — CHANGES_REQUESTED** [review]

File: references/proxy-template/pipeline.sh (hunk 11 above)
Comment (JoelGauci) [reply references/proxy-template/pipeline.sh]: ok done Daniel

**Review by danistrebel — APPROVED** [review]
LGTM 👍

**Pull request merged**